### ISSUE TITLE
How to get the hook while generating new tokens?

### ISSUE THREAD

**Issue by Meehaohao** [TransformerLensOrg/TransformerLens#697]
### Question

My prompt is "Which instrument does Henry Halstead mainly play? Please answer an instrument name. Answer: ", which is a question to LLM.
I want to get the cached hidden states of the _**LLM responsed tokens**_ while LLM is generating the response. How to do it?

On one hand, the code` logits, cache = model.run_with_cache(prompt, return_cache_object=True) ` only cache the hidden states of the prompt, because it dosen't run the generate function.

On the other hand, the code `output = model.generate(prompt, do_sample = False, max_new_tokens = 20)` only get the generated tokens or sentences, I can't get the Activation_cache of the generated Answer.

So how can I obtain the output answer and the Acativation_cache corresponding to the answer at the same time **_during one reasoning process_**?

**Comment by bryce13950** [comment]
Unfortunately, at this moment in time there is no integration of activation cache in the generate function. I don't see any reason why we can't add that as an option, but it would unfortunately be a pretty low priority given some other projects that are currently being worked on unless someone volunteers to do it.

**Comment by Meehaohao** [comment]
Got it, thank you.

**Comment by RecreationalMath** [comment]
I can add an opt-in flag (e.g. `return_cache`) to `TransformerBridge.generate()` that captures activations across the generated tokens by reusing the `run_with_cache` hook machinery (`names_filter` + cache hooks). Since `generate()` already runs a bridge forward each step, the hooks fire per token; the main thing to get right is the cache contract when `use_past_kv_cache=True` (each post-prompt step only sees the new token, so the `[batch, 1, ...]` slices get concatenated onto the prompt-step activations), which I'll document. I'll include a short usage example and a test in the same PR.

@jlarson4 happy to take this on, could you assign it to me?

**Comment by jlarson4** [comment]
@RecreationalMath you are now assigned, I look forward to reviewing your PR! Let me know if you run into any issues or questions

**Comment by RecreationalMath** [comment]
Some thoughts on the cache semantics before I open the PR. Refining my earlier sketch (which leaned toward Option A mentioned below): I prototyped two approaches and want to lay out the tradeoff, with some data.

- **(Option A) If we capture during the generation forwards:** with `use_past_kv_cache=True` (the default), each step only forwards the new token, so residual hooks come back as `[batch, 1, ...]` slices (concatenable) but `hook_pattern` comes back as ragged `[batch, heads, 1, k]` rows that can't be reassembled into the `[batch, heads, q, k]` matrix `run_with_cache` returns (and patterns aren't populated at all unless we force `output_attentions` each step). The final generated token never gets its own forward, so its activations are missing. And it runs on the cached eager-attention path, the same one behind the macOS-arm64 NaNs in #1322. Measured: on `distilgpt2` it matches a clean forward to ~1e-4, but on `pythia-70m` the attention output diverges by O(1) at the generated positions, so A would record activations that don't match a normal forward.
- **(Option B) If we recompute one clean forward over the full output:** `generate` would internally call `run_with_cache(out)` on the prompt+generated sequence and return its cache, so the result is consistent with the rest of TransformerLens, includes full attention patterns and all hook points, and holds prompt + response in one `ActivationCache`. It avoids the #1322 path. The cost is one extra forward, and for a causal LM that's numerically identical to a clean during-generation capture anyway (verified: 0.0 diff with KV cache off).

So B's only real tradeoff is one extra forward, in exchange for correctness, completeness, and consistency. I lean towards option B, where `generate` returns the same `ActivationCache` that `run_with_cache` would produce for the full prompt + generated sequence:
```python
out, cache = model.generate(prompt, max_new_tokens=20, return_cache=True)
# cache == run_with_cache(out)
```
Capturing during generation (option A) may still be worth having later as an explicit opt-in, but for a distinct purpose: the steering/intervention case, where hooks are active during generation and you want exactly the activations that ran. That path already works today via `with model.hooks(...)` around `generate`, so it isn't needed to resolve this issue.

Does Option B sound good, or do you specifically want the per-step generation-time activations? @jlarson4

**Comment by jlarson4** [comment]
@RecreationalMath I agree with Option B, thank you for the excellent analysis. The missing attention patterns and missing final-token activations would compound that into a feature that looks right, but returns silently-wrong caches.

A couple small implementation notes:

1. Plumb `names_filter` through `generate(return_cache=True, names_filter=...)` so users can scope the cache the same way they would on `run_with_cache`.
2. The cache for `prompt+max_new_tokens` can be large. Make sure to expose `device=` for cache offloading or note the memory implication in the docstring/example.
3. When `output_logits=True`, `generate` returns a ModelOutput-like object with `sequences` rather than a bare tensor. Option B's internal `run_with_cache(out)` will need to special-case this.

Looking forward to the PR!

**Comment by RecreationalMath** [comment]
FYI @jlarson4, this is completed by #1337 and can be closed.

**Comment by jlarson4** [comment]
Thanks for the tag @RecreationalMath, this is now closed!
